lib/src/codec/message_decoder.dart

```diff
@@ -0,0 +1,226 @@
+/*
+ * Package : Coap
+ * Author : S. Hamblett <steve.hamblett@linux.com>
+ * Date   : 14/05/2018
+ * Copyright :  S.Hamblett
+ */
+
+import 'package:typed_data/typed_data.dart';
+
+import '../coap_code.dart';
+import '../coap_constants.dart';
+import '../coap_empty_message.dart';
+import '../coap_message.dart';
+import '../coap_message_type.dart';
+import '../coap_option.dart';
+import '../coap_option_type.dart';
+import '../coap_request.dart';
+import '../coap_response.dart';
+import 'datagram/datagram_reader.dart';
+import 'specification.dart' as specification;
+
+/// Provides methods to parse incoming byte arrays to messages.
+class CoapMessageDecoder {
+  /// Instantiates.
+  CoapMessageDecoder(final Uint8Buffer data)
+      : reader = CoapDatagramReader(data) {
+    readProtocol();
+  }
+
+  /// The bytes reader
+  final CoapDatagramReader reader;
+
+  /// Checks if the decoding message is well formed.
+  bool get isWellFormed => version == specification.version;
+
+  int? _version;
+
+  /// Gets the version of the decoding message.
+  int? get version => _version;
+
+  int _tokenLength = 0;
+
+  int? _id;
+
+  /// Gets the id of the decoding message.
+  int? get id => _id;
+
+  /// Reads protocol headers.
+  void readProtocol() {
+    // Read headers
+    _version = reader.read(specification.versionBits);
+    _type = reader.read(specification.typeBits);
+    _tokenLength = reader.read(specification.tokenLengthBits);
+    _code = reader.read(specification.codeBits);
+    _id = reader.read(specification.idBits);
+  }
+
+  int? _type;
+
+  /// The type of the decoding message
+  int? get type => _type;
+
+  int? _code;
+
+  /// The code of the decoding message
+  int? get code => _code;
+
+  /// Parses the rest data other than protocol headers into the given message.
+  void parseMessage(final CoapMessage message) {
+    // Read token
+    if (_tokenLength > 0) {
+      message.token = reader.readBytes(_tokenLength);
+    } else {
+      message.token = CoapConstants.emptyToken;
+    }
+    // Read options
+    var currentOption = 0;
+    while (reader.bytesAvailable) {
+      final nextByte = reader.readNextByte();
+      if (nextByte == specification.payloadMarker) {
+        if (!reader.bytesAvailable) {
+          // The presence of a marker followed by a zero-length payload
+          // must be processed as a message format error
+          throw const FormatException('Marker followed by 0 length payload');
```

Comment:
In addition to the refactorings, I replaced the `StateError` here with a `FormatException` as otherwise, the error might crash an application if only exceptions are being caught.